cpp_sdk/cpp_dist_bundle/benchmarks/observation.cpp

```diff
@@ -0,0 +1,116 @@
+            << "Benchmark Type or Model, "
+            << "Measurement Taken, "
+            << "Measured Value, "
+            << "Batch Size, "
+            << "Number of Iterations"
+            << "\n";
+    }
+
+    for (const auto& observation : observations) {
+        out << observation << "\n";
+    }
+
+    out.close();
+}
+
+bool ObservationCSVWriter::doesFileExist(const std::string& path) {
+    std::ifstream f{m_path};
+    return f.good();
+}
+
+void appendObservationsFromTimes(const std::string &version, bool isGpuEnabled,
+                                 const std::string &benchmarkName, const std::string &benchmarkSubType,
+                                 const Parameters &params, std::vector<float> times,
+                                 ObservationList &observations) {
+    //
+    // times should be passed in NANOSECONDS as milliseconds do not have the resolution for
+    // some individual executions. So, the following divide through by 1000 to convert to ms
+    // for consumption down stream AFTER the calculations are done.
+    //
+    std::transform(times.begin(), times.end(), times.begin(), [&params](float val) {
+        return val / static_cast<float>(params.batchSize);
+    });
+    auto total = std::accumulate(times.begin(), times.end(), 0.0);
+    auto mean = total / params.numIterations;
+    const auto minmax = std::minmax_element(times.begin(), times.end());
+
+    const size_t sz = times.size();
+    auto variance_func = [&mean, &sz](float accumulator, const float &val) {
+        return accumulator + ((val - mean) * (val - mean) / (sz - 1));
+    };
+    auto variance = std::accumulate(times.begin(), times.end(), 0.0, variance_func) / 1000.f;
+
+    constexpr float nsPerMs{1000.f * 1000.f};
+    total /= nsPerMs;
+    mean /= nsPerMs;
+    variance /= nsPerMs * 1000.f;
+    *minmax.first /= nsPerMs;
+    *minmax.second /= nsPerMs;
+
+    // screen output for reporting progress to user
+    std::cout << "Average time " << benchmarkName;
+    if (!benchmarkSubType.empty()) {
+        std::cout << " (" << benchmarkSubType << ")";
+    }
+    auto precision{std::cout.precision()};
+    std::cout << ": " << std::fixed << std::setprecision(3) << mean << " ms | "
+        << std::defaultfloat << std::setprecision(precision)
+        << params.numIterations << " iterations " << std::endl;
+
+    observations.emplace_back(version, isGpuEnabled, benchmarkName, benchmarkSubType, "Total Time (ms)", params, total);
+    observations.emplace_back(version, isGpuEnabled, benchmarkName, benchmarkSubType, "Mean Time (ms)", params, mean);
+    observations.emplace_back(version, isGpuEnabled, benchmarkName, benchmarkSubType, "Variance (ms)", params, variance);
+    observations.emplace_back(version, isGpuEnabled, benchmarkName, benchmarkSubType, "Low (ms)", params, *minmax.first);
+    observations.emplace_back(version, isGpuEnabled, benchmarkName, benchmarkSubType, "High (ms)", params, *minmax.second);
+}
```

Comment:
Seems to me you'd only need to make changes here: https://github.com/getchui/sample-apps/blob/6a940d771864fc0423b12cab0efb5d1a4d3722ef/cpp_sdk/cpp_dist_bundle/benchmarks/observation.cpp#L51 and here https://github.com/getchui/sample-apps/blob/6a940d771864fc0423b12cab0efb5d1a4d3722ef/cpp_sdk/cpp_dist_bundle/benchmarks/observation.cpp#L28-L29 to accommodate the new columns. Is there more to it then that that I'm missing? Because if it's only that, I'd advise we take the column approach instead of 6 rows for each benchmark as having columns is much more readable. and less wasteful of space.